File: HashmapAndSet/largest_unique_number.py
from collections import Counter
from typing import List


class LargestUniqueNumber:
    def largestUniqueNumber(self, nums: List[int]) -> int:
        counts = Counter(nums)
        max_unique_num = -1
        for num, count in counts.items():
            if count == 1:
                max_unique_num = max(num, max_unique_num)

        return max_unique_num

    # Time complexity: O(n) where n is the number of elements in the input array.
    # We iterate through the input array once (O(n))
    # to count the occurrences of each number,
    # and then we iterate through the unique numbers
    # (O(n) in worst case where all numbers are unique)
    # to find the largest unique number.

File: HashmapAndSet/LargestUniqueNumber.java
import java.util.HashMap;
import java.util.Map;

public class LargestUniqueNumber {
  public int largestUniqueNumber(int[] nums) {
    Map<Integer, Integer> counts = new HashMap<>();
    for (int num : nums) {
      counts.merge(num, 1, Integer::sum);
    }

    int maxUniqueNum = -1;
    for (Map.Entry<Integer, Integer> count : counts.entrySet()) {
      if (count.getValue() == 1) {
        maxUniqueNum = Math.max(count.getKey(), maxUniqueNum);
      }
    }

    return maxUniqueNum;
  }
}

// Time Complexity: O(n) - We traverse the array once to count frequencies
// and then iterate through the map to find the largest unique number
// (O(n) in the worst case where all numbers are unique).
// Space Complexity: O(n) - In the worst case, all numbers are unique,
// and we store them in the map.